runtime/near-vm-logic/src/gas_counter.rs

```diff
@@ -49,48 +69,98 @@ impl GasCounter {
+                    min(self.max_gas_burnt, self.prepaid_gas - new_promises_gas);
+            }
+            self.fast_counter.burnt_gas = new_burnt_gas;
+            self.promises_gas = new_promises_gas;
             Ok(())
         } else {
-            use std::cmp::min;
-            let res = if new_burnt_gas > self.max_gas_burnt {
+            // if max_gas_burnt == prepaid_gas we must use GasExceeded error.
+            if new_burnt_gas > self.max_gas_burnt && self.max_gas_burnt != self.prepaid_gas {
+                self.fast_counter.burnt_gas = self.max_gas_burnt;
                 Err(HostError::GasLimitExceeded.into())
-            } else if new_used_gas > self.prepaid_gas {
-                Err(HostError::GasExceeded.into())
             } else {
-                unreachable!()
-            };
-
-            let max_burnt_gas = min(self.max_gas_burnt, self.prepaid_gas);
-            self.burnt_gas = min(new_burnt_gas, max_burnt_gas);
-            self.used_gas = min(new_used_gas, self.prepaid_gas);
+                self.fast_counter.burnt_gas = min(new_burnt_gas, self.prepaid_gas);
+                // Technically we shall do `self.promises_gas = 0;` or error paths, as in this case
+                // no promises will be kept, but that would mean protocol change.
+                // TODO: consider making this change!
+                assert!(self.prepaid_gas >= self.fast_counter.burnt_gas);
+                self.promises_gas = self.prepaid_gas - self.fast_counter.burnt_gas;
+                Err(HostError::GasExceeded.into())
+            }
+        }
+    }
 
-            res
+    // Optimized version of above function for cases where no promises involved.
+    pub fn burn_gas(&mut self, value: Gas) -> Result<()> {
+        let new_burnt_gas =
+            self.fast_counter.burnt_gas.checked_add(value).ok_or(HostError::IntegerOverflow)?;
+        if new_burnt_gas <= self.fast_counter.gas_limit {
+            self.fast_counter.burnt_gas = new_burnt_gas;
+            Ok(())
+        } else {
+            // if max_gas_burnt == prepaid_gas we must use GasExceeded error.
+            if new_burnt_gas > self.max_gas_burnt && self.max_gas_burnt != self.prepaid_gas {
+                self.fast_counter.burnt_gas = self.max_gas_burnt;
+                Err(HostError::GasLimitExceeded.into())
+            } else {
+                use std::cmp::min;
+                // Now we limit burnt gas with prepaid amount.
+                self.fast_counter.burnt_gas = min(new_burnt_gas, self.prepaid_gas);
+                // Technically we shall do `self.promises_gas = 0;` or error paths, as in this case
+                // no promises will be kept.
+                // TODO: consider making this change and fix tests/protocol!
+                assert!(self.prepaid_gas >= self.fast_counter.burnt_gas);
+                self.promises_gas = self.prepaid_gas - self.fast_counter.burnt_gas;
+                Err(HostError::GasExceeded.into())
```

Comment:
@olonho it seems that this part of the code is duplicated. Could we refactor it?

Reply:
Done